src/js/device-state.js

```diff
@@ -0,0 +1,94 @@
+import on from './on';
+import { add } from './class-list';
+import ownerWindow from './owner-window';
+import throttle from './throttle';
+
+const deviceStateClass = '.js-device-state';
+const regexWhiteSpace = /\s/g;
+const regexUnquote = /^['"]+|['"]+$/g;
+
+export default class DeviceStateObserver {
+  constructor() {
+    this.window = null;
+    this.lastContent = '';
+    this.hasStateChanged = false;
+  }
+
+  init() {
+    if (!this.node) {
+      this.node = document.body;
+      add(this.node, deviceStateClass);
+    }
+
+    const _handleResize = throttle(this.handleResize, 100);
+    this.offResize = on(ownerWindow(this.node), 'resize', _handleResize);
+    this.offOrientationchange = on(ownerWindow(this.node), 'orientationchange', _handleResize);
+  }
+
+  listen(callbackOnChange) {
+    this.node = document.querySelector(deviceStateClass);
+    this.callbackOnChange = callbackOnChange;
```

Comment:
correct.. more callbacks needs more instances. I wanted to be sure context is separate

Reply:
how does it handle first notification?
As I can see it only notifies on change, but what if the consumer needs the data at instanciation?

Reply:
1 observer for 1 listener for 1 callback

Reply:
first notifcation: u call triggerOnce method

Reply:
@AndyOGo line 39. Usage exmaple in src/components/m-datepicker/js/datepicker.js

Reply:
why

Reply:
May I ask which context you mean?
> correct.. more callbacks needs more instances. I wanted to be sure context is separate
context of `this`, or a contextual custom element, or some thing else?

Reply:
not `this` but the wcNode. So that there is no shared memory.

Reply:
relative to wcNode is always a new context. Ergo 4 items = 4 instances

Reply:
> why
Because I understand that, device state should:
- provide it's state upon consumption automatically (without the need to call `triggerOnce` from outside) - Since the whole concept of an observer is to get notified about subscription und not insist on being notified right now
- by attaching multiple `resize`/`orientationchange` events, those will be computed for every instance. By this it creates performance issues and memory issues. Overall this state should just be compute once and notify x subscribers
- Since `pub/sub` is generic pattern, I would favour one implementation so that devs hav to care only for one. If we have multiple implementation each dev has to first ask himself which oberserver do I need to import to observe x, y, z

Reply:
Alright, may I ask to see these?

> not `this` but the wcNode. So that there is no shared memory.
> relative to wcNode is always a new context. Ergo 4 custom elements = 4 instances

Reply:
@AndyOGo we either merge this PR or we forget about the datepicker. I dont have time to run behind nice to have or minimal milliseconds improvemts. If in future is needed, we can change it. But building the golden solution form beginning does not match CleanCode best practices. triggerOnce is there so that someone is not forced to render if he only wants to do it onchange. Is not a pub/sub pattern atm. Due to not common sharing of memory is impossible to implement what you suggest without hacking around window object.

Reply:
multiple listeners versus one that always call all subscriber and also cleanup logic of subscriber is irrelevant performance increase but relevant complexity increase

Reply:
No if we revert it wont work for datepicker and it trigger double on change due to the global shared listsner.

Reply:
agree DOM OPS are heavy.. so better render once!

Reply:
and how is this colerated to render once or not?

Reply:
it does not trigger more renderings

Reply:
it does render less

Reply:
hmh, maybe I miss some information here 🤔 
So the current implementation of `device-state` has bugs, I would be very interested to know them, because `<axa-footer-links>`, the switch between `<axa-header-navigation>` and `<axa-header-navigation-mobile>` and others rely on it.
And I thought it works properly and is well tested.

I would really appreciate to see with life code, what is broken or does not work as expected.

Reply:
i dont have time andy am so sorry. But the pub/sub is ONLY used in axa-footer-links

Reply:
True, `device-state` is only used for footer-links

Reply:
Though I would really like to see the bug...

Reply:
So, ISSUE 1

![bildschirmfoto 2018-09-05 um 17 24 31](https://user-images.githubusercontent.com/275235/45103607-9455d600-b130-11e8-8646-b3c12cf02fa3.png)

ISSUE 2
![bildschirmfoto 2018-09-05 um 17 25 30](https://user-images.githubusercontent.com/275235/45103676-b8b1b280-b130-11e8-8819-06433b8546f6.png)

![bildschirmfoto 2018-09-05 um 17 25 55](https://user-images.githubusercontent.com/275235/45103706-c7986500-b130-11e8-9d46-a556391bbc07.png)

ISSUE 3
When open and closing the links, the counters increases.

Reply:
Allright, so lets finish this.
Refactoring it is fine for me, though I would ask to refactor it into a singleton (maybe as follow-up). And I still would like to see what went wrong with the previous one...

Reply:
> So, ISSUE 1
True, this is the problem with our build not supporting code-splitting right now.
If I remember correctly we wanted to have one common chunks build for all components (so its a build issue not a `pub/sub` issue)

> ISSUE 2
Agreed that `4` count is weird 👍 

> ISSUE 3
I'm not sure which counters you mean, if those of `console.log('HERE')`. Agreed again, that sucks.